quaireaux/math/src/sha256.cairo

```diff
@@ -0,0 +1,321 @@
+    downcast(result).unwrap()
+}
+
+fn ssig0(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 7) | shl(x, 32 - 7)) & U32_MAX;
+    let x2 = (shr(x, 18) | shl(x, 32 - 18)) & U32_MAX;
+    let x3 = (shr(x, 3)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn ssig1(x: u32) -> u32 {
+    let x: u128 = upcast(x);
+    let x1 = (shr(x, 17) | shl(x, 32 - 17)) & U32_MAX;
+    let x2 = (shr(x, 19) | shl(x, 32 - 19)) & U32_MAX;
+    let x3 = (shr(x, 10)) & U32_MAX;
+    let result = x1 ^ x2 ^ x3;
+    downcast(result).unwrap()
+}
+
+fn sha256(mut data: Array<u8>) -> Array<u8> {
+    let u64_data_length: felt252 = (data.len() * 8).into();
+    let u64_data_length: felt252 = (data.len() * 8).into();
+    let u64_data_length: u128 = u64_data_length.try_into().unwrap();
+    let u64_data_length = u64_data_length & U64_MAX;
+
+    // add one
+    data.append(0x80);
+
+    add_padding(ref data);
+
+    // add length to the end
+    data.append(downcast(shr(u64_data_length, 56) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 48) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 40) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 32) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 24) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 16) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 8) & U8_MAX).unwrap());
+    data.append(downcast(shr(u64_data_length, 0) & U8_MAX).unwrap());
+
+    let u32_data_length = 16 * ((data.len() - 1) / 64 + 1);
+    let mut data = from_u8Array_to_u32Array(ref data, u32_data_length);
+    let mut h = get_h();
+    let mut k = get_k();
+    h = sha256_inner(ref data, 0, ref k, h);
+
+    from_u32Array_to_u8Array(ref h, 8)
+}
+
+fn from_u32Array_to_u8Array(ref data: Array<u32>, i: usize) -> Array<u8> {
+    quaireaux_utils::check_gas();
+    if i <= 0 {
+        return ArrayTrait::new();
+    }
+    let mut result = from_u32Array_to_u8Array(ref data, i - 1);
+    result.append(downcast(shr(upcast(*data[i - 1]), 24) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 16) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 8) & U8_MAX).unwrap());
+    result.append(downcast(shr(upcast(*data[i - 1]), 0) & U8_MAX).unwrap());
+    result
+}
+
+fn copy_array(ref from: Array<u32>, i: usize) -> Array<u32> {
+    quaireaux_utils::check_gas();
+    if i <= 0 {
+        return ArrayTrait::new();
+    }
+    let mut result = copy_array(ref from, i - 1);
+    result.append(*from[i - 1]);
+    result
+}
+
+fn sha256_inner(
+    ref data: Array<u32>, i: usize, ref k: Array<u32>, mut h: Array<u32>
+) -> Array<u32> {
+    quaireaux_utils::check_gas();
+    if 16 * i >= data.len() {
+        return h;
+    }
+    let mut w = create_w(ref data, i, 16);
+    create_message_schedule(ref w, 16);
+    let mut h2 = copy_array(ref h, h.len());
```

Comment:
If you:
```rust
use clone::Clone;
use array::ArrayTCloneImpl;
```
at the top, you can then delete the `fn copy_array` and replace this line by:
```suggestion
    let mut h2 = h.clone();
```